### ISSUE TITLE
Older revisions of Master OF can't be compiled anymore - missing libs zip in ci server?

### ISSUE THREAD

**Issue by armadillu** [openframeworks/openFrameworks#5903]
Trying to clone and run old projects that rely on older git revisions of OpenFrameworks master branch fails because libs that were hosted in ci.openframeworks.cc (for example http://ci.openframeworks.cc/libs/openFrameworksLibs_master_vs32.zip) are missing. 

Just to be clear, this happens when you clone from github and try to run the `OpenFrameworks\scripts\your_platform\download_libs-sh` script. The script fails to download some of the zip files.

It seems that the zip file has now been split across 4 zips (openFrameworksLibs_master_vs2015_32_[1..4].zip).

I can tell because luckily the internet archive crawled the OF ci libs site http://web.archive.org/web/20170725180057/http://ci.openframeworks.cc/libs/ 

Those libs are required to be able to compile projects against older versions of OpenFrameworks. I wonder if it would be possible to keep those files in http://ci.openframeworks.cc/libs/ as they were (hopefully with the same URL so scripts don't need to be modified), or at least let me get a copy of the files so I can host them somewhere else and update the old download scripts.

At the same time, I wonder if these zip file should be kept in a "versioned" context; I'm not sure how often the libs are updated, but I can't see the same version of these libs working for all of OpenFramework's past history?

**Comment by armadillu** [comment]
Maybe we could have all these zips in a git repo to keep track with actual revisions when they get changed. And make the scripts download them directly from github? Not sure how older scripts could point to the right version of the zip, but might be a step towards not losing the libs

**Comment by hiroMTB** [comment]
ci server only stores recent libs. 
Have you tried libs which came with official release zip?
Probabry it won't be nice idea to upload lib binary of all version apothecary build.

**Comment by armadillu** [comment]
The libs in the current zips seem to work, it's just that the `download_libs.sh` script at "OpenFrameworks/scripts/*/" fail for several platforms for this older revisions. We are using sort of an automated build system   , so this is quite a pain.

**Comment by hiroMTB** [comment]
Ah, I see your point.
Indeed it might be nice to store all libs of official release so that download_libs.sh does not fail in future.

Something like,
http://ci.openframeworks.cc/libs/master/***.zip
http://ci.openframeworks.cc/libs/v0.9.8/***.zip
http://ci.openframeworks.cc/libs/v0.9.7/***.zip
...

**Comment by hiroMTB** [comment]
For your quick solution, it might be nice to write custom script which direct libs on your server.

**Comment by armadillu** [comment]
For anyone else having this issue, this repo might help https://github.com/local-projects/OpenFrameworksLibFiles

**Comment by armadillu** [comment]
@arturoc now that we are moving into OF 0.10, is are there any plans to handle this? 

Specifically, people using the OF git repo as reference instead of working against the zip releases... 

I think we need some way of specifying what the libs download script downloads, other than blindly downloading zip files by name with no reference to what lib versions they contain.

Otherwise it's going to be a nightmare in the future to revisit/recompile old projects that are to be compiled against random revisions of master branch, trying to chase the right libs and headers. I think there would be value in storing revisions of the libs every time they are updated, not sure if we could sync it up with apothecary, or what versioning system to use... 

Maybe the lib folders could be handled as submodules or git-lfs / git annex?

**Comment by arturoc** [comment]
i'm not sure we should track every change in the binaries, my idea was to copy the binaries as they are every release the same way we create a tag on the repo so you can go back to a specific stable release but i think storing all the binaries would be overkill. 

If someone wants to go back to a specific commit hash i guess you can always go back to the corresponding apothecary commit and manually compile the libraries

**Comment by armadillu** [comment]
@hiroMTB the problem is that the last release is from Nov 2016, and I bet the libs have changed quite a few times between OF 0.9.8 and 0.10.0... If your projects are compiled against any revision between 0.9.8 and 0.10 you should be able to download/obtain the orignial libs for that revision. Just randomly trying libs from releases might work, but there's a huge risk of obscure bugs happening. 

I think you should be able to trace back any OF git revision from the past and recompile it with 100% guarantee that it will work as it did when someone checked it in, otherwise the versioning system is a bit useless.

In summary, I think the libs versions/releases can't be linked to OF release version numbering, as it happens more often that that.

**Comment by arturoc** [comment]
btw in linux for example most external dependencies are installed on the system and libraries work just fine from the really old versions in the ci servers (4 years old) to the latest in ubuntu 18.04 or archlinux so using a slightly old dependency shouldn't be such a big deal

**Comment by armadillu** [comment]
Maybe the Apothecary repo could have "releases" hosted in github as OF does, and the `download_libs.sh` could go fetch those instead of the files manually copied to `ci.openframworks.cc`

**Comment by arturoc** [comment]
I think the problem you are seeing happens because we've had to divide the builds in 3 and then in 4 "bundles" so the script has changed accordingly and if you run an old script it won't download all the libraries. We can probably make the script smarter and try to not break the format from now on so future scripts are always able to download old versions.

I'm almost sure that if you get the libraries from 0.9.8 add glm to them and put them in the right folders in 0.10 they would work without problem in most platforms. So the problem is not really with the libraries but with the packages.

Right now if you want to work with something between 0.9.8 and 0.10 i think the safest would be to checkout master, run the latest download_libs, checkout the hash you want to use and it should work fine.

I'll take a look at making the script smarter so it can download any number of files and add parameters to download a specific version once we have a stable release for 0.10. In any case master is considered unstable so if you work with it you are exposed to some breaking changes.

**Comment by hiroMTB** [comment]
It's a Apothecary commit driven. Here is build script, upload line is on the bottom.
https://github.com/openframeworks/apothecary/blob/master/scripts/build.sh

Might be nice to make log file on /libs directory which indicate commit hash of Apothecary?
To make it easier to generate libs by ourselves.

**Comment by armadillu** [comment]
@arturo yes in theory, but in practive I've had problems with poco headers changing and breaking compilation across different versions. I can't use the current `download_libs.sh` in older revisions because of this (poco breaks), I had to resort to older versions of the libs "zips"... Which is why I'm hosting them in the repo mentioned above (https://github.com/local-projects/OpenFrameworksLibFiles).

Yep I agree Master is meant to be unstable, but if it doesn't take a lot of effort to keep commits reproducible (ie so that you get the libs that go with that revision), why not? 

The real issue is that at the company I work for we don't use OF "zip" releases, we reference the github repo so that we can have OF itself as a submodule of the project (together with any extra addons that are used). This makes cloning and changing revisions of OF / addons very flexible, but trouble came when the libs were pushed out of the OF repo (which is a smart decision). At this point, you have to trust the download script to work for projects to be "compilable" again, and it doesn't anymore.

@hiroMTB this is great info, I might look into that... Or even have some bot automatically checking for new revisions of apothecary and making a release of all the libs and copying them some public  host with a simple folder structure similar to `http://libs.openframeworks.cc/apothecary_repo_git_sha1/platformX_architectureX_.zip` so that the download script can get those instead

**Comment by arturoc** [comment]
Can you post the errors you are having with poco? as i said we use poco versions (and most of the libraries really)  4 years old and the very latest with the same codebase in linux and it works fine.

I understand that your company might have this specific need but the master branch of this repo is not supposed to be used for production but to develop OF itself and we don't have the time or resources to make it into an always stable branch (i doubt any oss project does that), you can still use OF from github but use the latest stable. If your company needs something more specific then it might need to put the resources to solve this issue (if you do so it would be nice to make it publicly available :)  but for us at the moment is really not a priority.

**Comment by hiroMTB** [comment]
@armadillu 
How about proposing PR like following feature?

- oF ci script which does not download libs but compiles.
- oF download Apothecary repo (which is synchronized to each oF commit )
- Auto run Apothecary's build script and generate libs

Once I run Apothecary on my OSX without Docker and it works nicely.

**Comment by arturoc** [comment]
yes that's what i was suggesting before if you really need something between releases you can just download the corresponding commit in apothecary and build the libraries from there. apothecary is already a submodule and although we might forget to sync it every time there's a change, updating the submodules should give you something very near to what it was or you can just look for a specific commit by date.

**Comment by hiroMTB** [comment]
> apothecary is already a submodule

Ah true! 
Then it will be no problem. Just run Apothecary build script instead of oF download script.

**Comment by armadillu** [comment]
Thx for the feedback guys, totally valid points. I will try find an apothecary worklfow that work for us and let you know! 

I will also post the Poco errors we've been seeing.

**Comment by armadillu** [comment]
Here's one of the Poco errors I've seen;

```
1>------ Build started: Project: mySketch, Configuration: Debug x64 ------
1>     Creating library bin\mySketch_debug.lib and object bin\mySketch_debug.exp
1>PocoFoundationmdd.lib(Environment.obj) : error LNK2019: unresolved external symbol GetAdaptersInfo referenced in function "public: static void __cdecl Poco::EnvironmentImpl::nodeIdImpl(unsigned char (&)[6])" (?nodeIdImpl@EnvironmentImpl@Poco@@SAXAEAY05E@Z)
1>bin\mySketch_debug.exe : fatal error LNK1120: 1 unresolved externals
```

This happens when trying to compile and older revision of OF with the current libs (Running the newest download_libs.sh script). This is on Windows 10 - VS 2015, an empty project that just includes ofxPoco.

When running the modified script that fetches older versions of the libs (found in here https://github.com/local-projects/OpenFrameworksLibFiles) the same exact project builds fine.

Also, when the project actually uses Poco code, I've also seen this:

```
1>------ Build started: Project: mySketch, Configuration: Debug x64 ------
1>LINK : fatal error LNK1104: cannot open file 'libcrypto.lib'
========== Build: 0 succeeded, 1 failed, 1 up-to-date, 0 skipped ==========
```

I'm guessing this is related to OpenSSL, but I'm not sure how needs code from that lib, as it's not currently included inside the OpenSSL libs that come with OF nor poco.

**Comment by arturoc** [comment]
this is because newest poco needs a new library from the system in windows, if you copy the addon_config.mk from the newest ofxPoco it should work without problem

**Comment by armadillu** [comment]
ok makes sense, let me try that...

It doesn't seem to work. 
```
1>  openframeworksLib.vcxproj -> C:\Users\oriol\Desktop\lp-internal.OF_Project_template\OpenFrameworks\libs\openFrameworksCompiled\project\vs\..\..\lib\vs\x64\openframeworksLib_debug.lib
2>------ Build started: Project: mySketch, Configuration: Debug x64 ------
2>LINK : fatal error LNK1104: cannot open file 'libcrypto.lib'
```
I think the poco `addon_config.mk` is missing the crypto libs?
https://github.com/openframeworks/openFrameworks/blob/master/addons/ofxPoco/addon_config.mk#L180-L182

**Comment by arturoc** [comment]
does it work on current master? i think in windows we include those libraries in the core cause curl needs them

**Comment by armadillu** [comment]
I was just testing on the 0.10 RC1, getting VS 2017 installed right now... will test Master next

**Comment by arturoc** [comment]
yeah RC1 should be the same as master

just checked the libraries in current master and openssl is in libs for vs but it doesn't include libcrypto anymore, but it's not in your libaries packages either. Is this happening with an old project? probably running the PG over it will solve the issue

**Comment by arturoc** [comment]
can you tell me which OF commit are you using when it gives you that error? i'm not sure what you mean by using RC1 are you using the libraries from RC1 or an old project with RC1

**Comment by armadillu** [comment]
I thought that was the issue, so I re-created the project with the PG.exe from 0.10RC1. compiling a project with ofxPoco alone is fine, but adding ofxSimpleHttp in it raises the libcrypto issue.

BTW 0.10rc1 just finished compiling and it reports the same; missing libcrypto.lib. This is testing on a brand new project created with the PG.exe from 0.10rc1, just including ofxPoco and ofxSimpleHttp.

Note that if I remove ofxSimpleHttp from the project, it compiles fine. 

Also, somewhat unrelated, 0.10rc1 with vs2017 throws a lot of `Unknown compiler version - please run the configure tests and report the results` during compilation.

**Comment by armadillu** [comment]
I'm using a zip straight from the forum with 0.10RC1. Not mixing in anything at this point, only adding ofxSimpleHttp addon.

**Comment by arturoc** [comment]
ah ok, can you open a different issue? i was thinking you were speaking about the original problem stll

**Comment by hiroMTB** [comment]
@armadillu 
Lets close this issue?

**Comment by armadillu** [comment]
I see that http://ci.openframeworks.cc/libs has now a directory named `0.10.0`, with updated libs (created May 7th) while the libs at the "root" are pre 0.10.0 (creation date is March 23rd).

But it seems that current master (0.10.0+) download_libs.sh script still gets the old libs,

```
./download_libs.sh 
Downloading openFrameworksLibs_master_osx1.tar.bz2
--2018-06-19 11:42:14--  http://ci.openframeworks.cc/libs/openFrameworksLibs_master_osx1.tar.bz2
...
```

I would expect it to get the files from `http://ci.openframeworks.cc/libs/0.10.0/` instead? 
Is there any documentation on how this will be organized from now on?

**Comment by armadillu** [comment]
@arturoc I think this can be closed now?

**Comment by arturoc** [comment]
yes thanks!
